### DISCUSSION TITLE
Uncached Fixtures?

### DISCUSSION THREAD

**Discussion by ax3l** [pytest-dev/pytest#10387] [Q&A]
Is there a way to deactivate caching in a fixture in pytest? Or is there a recommended way to write generators of objects that are uncached in pytest but still use decorators like `parametrize`?

This is based a somewhat on the original question on behavior of caches in https://github.com/pytest-dev/pytest/issues/5642#issuecomment-1279612764 and on how to use generators across tests that should yield objects that shall not be cached.

**Comment by ax3l** [comment]
The work-around I came up with for now, to keep all using the nice parametrizations in pytest, is to make the fixture return a generator (which can be cached w/o problems for my use case).

Original:
```py
@pytest.fixture(scope="function", params=list(itertools.product([1, 3], [0, 1])))
def obj(C, D, request):
    A = request.param[0]
    B = request.param[1]
    obj = mymodule.Obj(A, B, C, D)
    return obj  # don't want to cache this between tests
```

Uncached fixture:
```py
@pytest.fixture(scope="function", params=list(itertools.product([1, 3], [0, 1])))
def make_obj(C, D, request):

    def create():
        A = request.param[0]
        B = request.param[1]
        obj = mymodule.Obj(A, B, C, D)
        return obj

    return create
```
(in the test, the first thing I call is `obj = make_obj()`.

**Comment by RonnyPfannschmidt** [comment]
It's absolutely unclear what you mean

Fixtures in function scope are not cached

And if you want to rely on `__del__` for cleanup, ideally don't, it's absolutely broken

**Comment by ax3l** [comment]
> It's absolutely unclear what you mean

Did you have a chance to see the example in the link in the description?
https://github.com/pytest-dev/pytest/issues/5642#issuecomment-1279612764
```py
# conftest.py
import gc
import sys
import numpy as np

import pytest


@pytest.fixture(scope="function")  # can I disable pytest caching for this yield?
def obj():
    v = np.array([1, 2, 3])
    print("startup: ", sys.getrefcount(v) - 1)  # refs: 1
    yield v
    gc.collect()
    print("\nteardown: ", sys.getrefcount(v) - 1)  # refs: 3 - why?
    del v  # still refs remaining
```
```py
# test_foo.py
import sys

def test_foo(obj):
    print("++ test_foo ++")
    print("in test:", sys.getrefcount(obj) - 1)
```

> Fixtures in function scope are not cached

Oh really? That's confusing to me, because in the linked example there are still 3 references (in CPython) on the function-scope yielded object:
```
startup:  1
++ test_foo ++
in test: 6
PASSED
teardown:  3
```

> And if you want to rely on `__del__` for cleanup, ideally don't, it's absolutely broken

Can you please be a bit more detailed what you mean by that?

**Reply by RonnyPfannschmidt** [reply]
the reference count cmoes from the fact that 
https://github.com/pytest-dev/pytest/blob/3dac833a520cc299c20b7270db87d62412b4119a/src/_pytest/fixtures.py#L1023-L1044
manages one of them an remvoes it **after** teardown,
the other reference comes from 

https://github.com/pytest-dev/pytest/blob/3dac833a520cc299c20b7270db87d62412b4119a/src/_pytest/fixtures.py#L544-L549
preparing the parameters to pass over

storing references to the objects in the fixture lifecycle is a necessity to pass over the correct values at the correct point in time

a starting point for `__del__` is https://docs.python.org/3/reference/datamodel.html#object.__del__

in practical terms - its demonstrably wrong to try to rly on `__del__` for resource cleanup, its at best used for memory cleanup, and cant by design support clean resource cleanup

clean resource cleanup is best done with context managers

**Comment by stevenmanton** [comment]
Thanks for this discussion! The issue of the reference count is very sneaky and can cause some very subtle issues, particularly for people trying to control garbage collection. For example, this comes up often when testing multiple models that run on a GPU. This simple strategy does not work:

```python
@pytest.fixture(scope="class")
def model():
    yield load_model()
    gc.collect()
    torch.cuda.empty_cache()
```

At first glance, one would expect the garbage collection to remove all references to the model and thus allow torch to free up the GPU memory. However, as @RonnyPfannschmidt pointed out, there are still references to the model! This means that if you run two GPU models back-to-back, Python might not have garbage-collected (and thus allowed torch to clear the GPU memory) by the time the second model loads. (See, for example, this [discussion](https://github.com/pytest-dev/pytest/discussions/10296)).

There seem to be two approaches that do work. First, you can garbage-collect _before_ loading each model. This deletes all references to tensors on the GPU and that allows torch to reclaim that memory. For example:

```python
# Approach 1:
@pytest.fixture(scope="class")
def model():
    gc.collect()
    yield load_model()
```

The second approach would be to delete all the tensors (or objects containing tensors) from the model and _then_ garbage-collect. For example:

```python
# Approach 2:
@pytest.fixture(scope="class")
def model():
    model = load_model()
    yield model
    del model.text_model
    del model.vision_model
    gc.collect()
```

Approach 2 has the advantage of clearing the memory after each test, though it's more boilerplate and it depends on each model.

**Reply by RonnyPfannschmidt** [reply]
Interesting find, this ought to become a contextmanager ideally dupplied bz torch
